pkg/controller/clusterconnection/clusterconnection_controller.go

```diff
@@ -323,6 +335,67 @@ func (r *ReconcileConnection) Reconcile(ctx context.Context, request reconcile.R
 		trustedCertBundle.AddCertificates(secret)
 	}
 
+	// Determine the current deployment availability.
+	var currentAvailabilityTransition metav1.Time
+	var currentlyAvailable bool
+	guardianDeployment := v1.Deployment{}
+	err = r.Client.Get(ctx, client.ObjectKey{Name: render.GuardianDeploymentName, Namespace: render.GuardianNamespace}, &guardianDeployment)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		r.status.SetDegraded(operatorv1.ResourceReadError, "Failed to read the deployment status of Guardian", err, reqLogger)
+		return reconcile.Result{}, nil
+	} else if err == nil {
+		for _, condition := range guardianDeployment.Status.Conditions {
+			if condition.Type == v1.DeploymentAvailable {
+				currentAvailabilityTransition = condition.LastTransitionTime
+				if condition.Status == corev1.ConditionTrue {
+					currentlyAvailable = true
+				}
+				break
+			}
+		}
+	}
+
+	// If the deployment availability has changed and is currently available, we update the resolved proxy configuration.
+	// We only update the resolved proxy configuration in this scenario (rather than every reconcile) to limit the number
+	// of pod queries we make.
+	if !currentAvailabilityTransition.Equal(&r.lastAvailabilityTransition) && currentlyAvailable {
+		// Query guardian pods.
+		labelSelector := labels.SelectorFromSet(map[string]string{
+			"app.kubernetes.io/name": render.GuardianDeploymentName,
+		})
+		pods := corev1.PodList{}
+		err := r.Client.List(ctx, &pods, &client.ListOptions{
+			LabelSelector: labelSelector,
+			Namespace:     render.GuardianNamespace,
+		})
+		if err != nil {
+			r.status.SetDegraded(operatorv1.ResourceReadError, "Failed to list the pods of the Guardian deployment", err, reqLogger)
+			return reconcile.Result{}, nil
+		}
+
+		// Parse the pod spec to resolve the proxy config.
+		var proxyConfig *httpproxy.Config
+		for _, pod := range pods.Items {
```

Comment:
Since we're already considering it viable that the operator might have a different proxy address than the guardian pods (which is why we're getting the proxy address directly from the guardian pod), it would probably be equally viable that the guardian pods have different proxy addresses from each other (in the event that we did allow multiple guardian pods, which would be a good idea to do eventually).

It probably wouldn't be much more work to collect a list of proxy configs, and apply them all to the policy.

I think with that addition it makes the solution fairly bullet proof.